### ISSUE TITLE
spurious, vanishing subgoals

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#3443]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 3443
Date: 2014-07-20 23:08:42 +0200
From: Kevin Hamlen &lt;<hamlen@utdallas.edu>&gt;
To:   
CC:   @_aspiwack, coq-bugs-redist@lists.gforge.inria.fr

Last updated: 2014-07-22 11:31:37 +0200

## Bugzilla Comment ID: 7294
Date: 2014-07-20 23:08:42 +0200
From: Kevin Hamlen &lt;<hamlen@utdallas.edu>&gt;

On Coq 8.4pl2 (PC Windows), the following code yields spurious "shadow" subgoals that mysteriously vanish when the subgoals they shadow are solved.

Tactic Notation "twice" tactic(T) := T;T.

Tactic Notation "mycases" hyp(H) :=
  twice (match False with _ => destruct H as [H|H] end).

Goal True\/False -> True.
  intro H.
  mycases H.
  (* Why did the above succeed? Shouldn't the second destruct fail?
     Why did it yield four subgoals instead of two? *)
  exact I.
  (* Whoa! The second subgoal vanished by solving the first! *)
  exact I.
  (* The last subgoal vanished by solving its predecessor! *)
Qed.

## Bugzilla Comment ID: 7302
Date: 2014-07-21 13:36:44 +0200
From: @_aspiwack

Ok. This is quite funny.

TL;DR: you can fix all your troubles by defining mycases as:

  Tactic Notation "mycases" hyp(H) :=
    twice (idtac;match False with _ => destruct H as [H|H] end).

The heart of the problem is the semantics of match in Ltac: it is not a tactic to be executed when toplevel is reached, but an immediate computation (prepending [idtac] as above prevents the computation of [match …]).

In particular,

  [twice (match False with _ => destruct H as [H|H] end)]

is not equivalent to

  [match False with _ => destruct H as [H|H] end;
   match False with _ => destruct H as [H|H] end]

It is rather something like

  [let x := match False with _ => destruct H as [H|H] end in x;x]

Even though you cannot write that in Ltac in v8.4 (you'll get an error), this is how it behaves internally.

The question is: what does [x] do. Well, in v8.4, it would seem that it tells the proof engine to just append a list of goals to the available goals. Duplicating [x] as you did by a clever use of tactic notations adds the goals twice (without executing destruct H twice however). Fortunately v8.4 tracks solved goals so the duplicate goals are solved when one of their copy is (you got it slightly wrong, by the way, solving the first goal also solves the third goal not the second one). The v8.3 version seems to simply fail with an assertion failure (I'm not going to track this one down, though).

The trunk version of Coq does not exhibit that behaviour because, there, [x] just means "do nothing". So, in trunk, you would get one destruct but the correct number of goals. Since it is a very specific bug, and that solving it would be pretty hard, I will not be fixing it for the 8.4, and consider this bug solved.

## Bugzilla Comment ID: 7316
Date: 2014-07-21 18:11:42 +0200
From: Kevin Hamlen &lt;<hamlen@utdallas.edu>&gt;

Thanks for the analysis, Arnaud.

Just to clarify: I'm aware that the match is evaluating too early and how to avoid that problem, but I constructed this (intentionally flawed) code in order to build a minimal working example of the bug.

It's good to hear that it's already fixed in the trunk.  Even though the minimal working example I submitted is not very useful, in practice I have found this bug to manifest in more problematic ways.  For example, my first exposure to it completely crashed coqtop because something inside a repeat-loop was introducing extra subgoals that it shouldn't.  The result was an errorless crash, presumably due to resource exhaustion---an unpleasant outcome that was rather difficult to track down!

## Bugzilla Comment ID: 7325
Date: 2014-07-22 11:31:37 +0200
From: @_aspiwack

Unpleasant indeed. I'm afraid, however, that the issue of effectively debugging Ltac code is still an open one.
